subsys/bluetooth/services/cts_client.c

```diff
@@ -0,0 +1,222 @@
+	time->exact_time_256.year = sys_get_le16(data);
+	data += sizeof(uint16_t);
+	time->exact_time_256.month = *data++;
+	time->exact_time_256.day = *data++;
+	time->exact_time_256.hours = *data++;
+	time->exact_time_256.minutes = *data++;
+	time->exact_time_256.seconds = *data++;
+
+	/* Day of week. */
+	time->exact_time_256.day_of_week = *data++;
+
+	/* Fractions of a second. */
+	time->exact_time_256.fractions256 = *data++;
+
+	/* Reason for updating the time. */
+	time->adjust_reason.manual_time_update = *data & BIT(0);
+	time->adjust_reason.external_reference_time_update =
+		(*data & BIT(1)) ? 1 : 0;
+	time->adjust_reason.change_of_time_zone = (*data & BIT(2)) ? 1 : 0;
+	time->adjust_reason.change_of_daylight_savings_time =
+		(*data & BIT(3)) ? 1 : 0;
+
+	return 0;
+}
+
+/**@brief Function for sanity check of a Current Time structure.
+ *
+ * @param[in] time  Current Time structure.
+ *
+ * @retval 0        If the time struct is valid.
+ * @retval -EINVAL  If the time is out of bounds.
+ */
+static int current_time_validate(struct bt_cts_current_time_char *time)
+{
+	if ((time->exact_time_256.year > CTS_YEAR_MAX) ||
+	    ((time->exact_time_256.year < CTS_YEAR_MIN) &&
+	     (time->exact_time_256.year != 0)) ||
+	    time->exact_time_256.month > 12 || time->exact_time_256.day > 31 ||
+	    time->exact_time_256.hours > 23 ||
+	    time->exact_time_256.minutes > 59 ||
+	    time->exact_time_256.seconds > 59 ||
+	    time->exact_time_256.day_of_week > 7) {
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+/**@brief Function for handling of Current Time read response.
+ *
+ * @param[in] conn    Connection object.
+ * @param[in] err     ATT error code.
+ * @param[in] params  Read parameters used.
+ * @param[in] data    Pointer to the data buffer.
+ * @param[in] length  The size of the received data.
+ *
+ * @retval BT_GATT_ITER_STOP  Stop reading
+ */
+static uint8_t bt_cts_client_read_callback(struct bt_conn *conn, uint8_t err,
+					   struct bt_gatt_read_params *params,
+					   const void *data, uint16_t length)
+{
+	struct bt_cts_client *cts_c;
+	struct bt_cts_client_evt evt;
+	int err_code;
+
+	/* Retrieve CTS client module context. */
+	cts_c = CONTAINER_OF(params, struct bt_cts_client, read_params);
+
+	if (!err) {
+		err_code = current_time_decode(&evt.current_time, data, length);
+		if (err_code) {
+			/* The data length was invalid. Decoding was not completed. */
+			evt.evt_type = BT_CTS_CLIENT_EVT_INVALID_TIME;
+		} else {
+			/* Verify that the time is valid. */
+			err_code = current_time_validate(&evt.current_time);
+			if (err_code) {
+				/* Invalid time received. */
+				evt.evt_type = BT_CTS_CLIENT_EVT_INVALID_TIME;
+			} else {
+				/* Valid time received. */
+				evt.evt_type = BT_CTS_CLIENT_EVT_CURRENT_TIME;
+			}
+		}
+		cts_c->evt_handler(cts_c, &evt);
+	}
+
+	k_sem_give(&cts_c->read_params_sem);
+
+	return BT_GATT_ITER_STOP;
+}
+
+int bt_cts_client_current_time_read(struct bt_cts_client *cts_c,
+				    k_timeout_t timeout)
```

Comment:
Did you base this on `bt_hogp_pm_update`?

Usually for these read functions we return -EBUSY instead of queueing using a semaphore.
I would prefer to do that here as well (see bt_hogp_rep_read and bt_bas_read_battery_level for example).

Reply:
Will do to align with other services.

This CTS PR is migrated from nRF5 SDK. I had not given a lot of thought to the design. My initial goal is to make a functional sample.